### PR TITLE
ARROW-16382: [Python] Disable memory mapping by default in pyarrow

### PR DESCRIPTION
**Pull request by AlvinJ15**
[Python] Disable memory mapping by default in pyarrow

### CODE REVIEW

File: python/pyarrow/ipc.pxi [hunk 1]
```diff
@@ -399,21 +399,23 @@ cdef class MessageReader(_Weakrefable):
                         "instead.".format(self.__class__.__name__))
 
     @staticmethod
-    def open_stream(source):
+    def open_stream(source, use_memory_map=False):
```
Comment (amol-) [inline python/pyarrow/ipc.pxi:402]: I think `user_memory_map` here is misleading for a user.
`source` can be a `PythonFile` or `NativeFile`, in such case the memory map option won't do anything. The option is only effective if the `source` is a string.

Given that the user can pass a `MemoryMappedFile` if they want to explicitly use `memory_map`. I think we should remove the option and explicitly document to use `MemoryMappedFile` as the source if you want to use memory mapping.

Comment (amol-) [reply python/pyarrow/ipc.pxi:402]: Same applies for the other cases where `_get_input_stream` is used and a `use_memory_map` option was added.

File: python/pyarrow/ipc.pxi [hunk 2]
```diff
@@ -839,15 +840,15 @@ cdef class _RecordBatchFileReader(_Weakrefable):
...
+              MemoryPool memory_pool=None, use_memory_map=False):
         self.options = options.c_options
         self.options.memory_pool = maybe_unbox_memory_pool(memory_pool)
         try:
             source = as_buffer(source)
         except TypeError:
             pass
 
-        get_reader(source, True, &self.file)
+        get_reader(source, use_memory_map, &self.file)
```
Comment (amol-) [inline python/pyarrow/ipc.pxi:851]: As previous cases it should support `MemoryMappedFile`

```suggestion
        get_reader(source, False, &self.file)
```

File: python/pyarrow/ipc.pxi [hunk 3]
```diff
@@ -1089,7 +1092,7 @@ def read_schema(obj, DictionaryMemo dictionary_memo=None):
     if isinstance(obj, Message):
         raise NotImplementedError(type(obj))
 
-    get_reader(obj, True, &cpp_file)
+    get_reader(obj, use_memory_map, &cpp_file)
```
Comment (amol-) [inline python/pyarrow/ipc.pxi:1095]:
```suggestion
    get_reader(obj, False, &cpp_file)
```

File: python/pyarrow/ipc.pxi [hunk 4]
```diff
@@ -1065,7 +1066,7 @@ def read_message(source):
     return result
 
 
-def read_schema(obj, DictionaryMemo dictionary_memo=None):
+def read_schema(obj, DictionaryMemo dictionary_memo=None, use_memory_map=False):
```
Comment (amol-) [inline python/pyarrow/ipc.pxi:1069]:
```suggestion
def read_schema(obj, DictionaryMemo dictionary_memo=None):
```

This is expected to work on `buffer` or `Message`, none of them can be memory_mapped

File: python/pyarrow/ipc.pxi [hunk 5]
```diff
@@ -1075,6 +1076,8 @@ def read_schema(obj, DictionaryMemo dictionary_memo=None):
     dictionary_memo : DictionaryMemo, optional
         Needed to be able to reconstruct dictionary-encoded fields
         with read_record_batch
+    use_memory_map : boolean, default False
+        Use memory mapping when opening file on disk
```
Comment (amol-) [inline python/pyarrow/ipc.pxi:1080]:
```suggestion
```

File: python/pyarrow/ipc.py [hunk 6]
```diff
@@ -44,11 +44,16 @@ class RecordBatchStreamReader(lib._RecordBatchStreamReader):
...
     memory_pool : MemoryPool, default None
         If None, default memory pool is used.
+    use_memory_map : bool, default False
+        If the source is a file path, use a memory map to read file, which can
+        improve performance in some environments.
     """
 
-    def __init__(self, source, *, options=None, memory_pool=None):
+    def __init__(self, source, *, options=None, memory_pool=None,
+                 use_memory_map=False):
```
Comment (amol-) [inline python/pyarrow/ipc.py:53]:
```suggestion
    def __init__(self, source, *, options=None, memory_pool=None):
```

File: python/pyarrow/serialization.pxi [hunk 7]
```diff
@@ -550,7 +555,8 @@ def deserialize(obj, SerializationContext context=None):
     return _deserialize(obj, context=context)
 
 
-def _deserialize(obj, SerializationContext context=None):
+def _deserialize(obj, SerializationContext context=None, use_memory_map=False):
```
Comment (amol-) [inline python/pyarrow/serialization.pxi:558]:
```suggestion
def _deserialize(obj, SerializationContext context=None):
```

File: python/pyarrow/serialization.pxi [hunk 8]
```diff
@@ -550,7 +555,8 @@ def deserialize(obj, SerializationContext context=None):
...
     source = BufferReader(obj)
-    serialized = _read_serialized(source, base=obj)
+    serialized = _read_serialized(
+        source, base=obj, use_memory_map=use_memory_map)
```
Comment (amol-) [inline python/pyarrow/serialization.pxi:561]:
```suggestion
    serialized = _read_serialized(source, base=obj)
```

File: python/pyarrow/ipc.pxi (hunk 1 above)
Comment (amol-) [inline python/pyarrow/ipc.pxi]:
```suggestion
    def open_stream(source):
```

File: python/pyarrow/ipc.pxi [hunk 9]
```diff
@@ -399,21 +399,23 @@ cdef class MessageReader(_Weakrefable):
...
+    def open_stream(source, use_memory_map=False):
         """
         Open stream from source.
 
         Parameters
         ----------
         source
             A readable source, like an InputStream
+        use_memory_map : boolean, default False
+            Use memory mapping when opening file on disk
```
Comment (amol-) [inline python/pyarrow/ipc.pxi]:
```suggestion
```

File: python/pyarrow/ipc.pxi [hunk 10]
```diff
@@ -562,14 +564,14 @@ cdef class _RecordBatchStreamWriter(_CRecordBatchWriter):
                                  self.options))
 
 
-cdef _get_input_stream(object source, shared_ptr[CInputStream]* out):
+cdef _get_input_stream(object source, shared_ptr[CInputStream]* out, use_memory_map=False):
```
Comment (amol-) [inline python/pyarrow/ipc.pxi]:
```suggestion
cdef _get_input_stream(object source, shared_ptr[CInputStream]* out):
```

File: python/pyarrow/ipc.pxi [hunk 11]
```diff
@@ -562,14 +564,14 @@ cdef class _RecordBatchStreamWriter(_CRecordBatchWriter):
...
     try:
         source = as_buffer(source)
     except TypeError:
         # Non-buffer-like
         pass
 
-    get_input_stream(source, True, out)
+    get_input_stream(source, use_memory_map, out)
```
Comment (amol-) [inline python/pyarrow/ipc.pxi]:
```suggestion
    get_input_stream(source, False, out)
```

File: python/pyarrow/ipc.pxi (hunk 9 above)
Comment (amol-) [reply python/pyarrow/ipc.pxi]: Let's add to the `source` docstring that you can pass `MemoryMappedFile` if you want to use memory mapping.

**Review by amol- — CHANGES_REQUESTED** [review]

**Review by amol- — APPROVED** [review]

**Comment by AlvinJ15** [comment]
@ursabot please benchmark

**Pull request merged**